data-prepper-plugins/otlp-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/otlp/buffer/OtlpSinkBuffer.java

```diff
@@ -0,0 +1,209 @@
+     * @param sinkMetrics the metrics collector to use
+     */
+    public OtlpSinkBuffer(@Nonnull final OtlpSinkConfig config, @Nonnull final OtlpSinkMetrics sinkMetrics) {
+        this(config, sinkMetrics, null, null);
+    }
+
+    /**
+     * Visible for testing only: constructs an OTLP sink buffer with injected encoder and sender.
+     *
+     * @param config      the OTLP sink configuration
+     * @param sinkMetrics the metrics collector
+     * @param encoder     custom OTLP encoder (or null to use default)
+     * @param sender      custom HTTP sender (or null to use default)
+     */
+    @VisibleForTesting
+    OtlpSinkBuffer(@Nonnull final OtlpSinkConfig config,
+                   @Nonnull final OtlpSinkMetrics sinkMetrics,
+                   final OTelProtoStandardCodec.OTelProtoEncoder encoder,
+                   final OtlpHttpSender sender) {
+
+        this.sinkMetrics = sinkMetrics;
+        this.encoder = encoder != null ? encoder : new OTelProtoStandardCodec.OTelProtoEncoder();
+        this.sender = sender != null ? sender : new OtlpHttpSender(config, sinkMetrics);
+
+        this.maxEvents = config.getMaxEvents();
+        this.maxBatchBytes = config.getMaxBatchSize();
+        this.flushTimeoutMillis = config.getFlushTimeoutMillis();
+
+        this.queue = new LinkedBlockingQueue<>(getQueueCapacity());
+        sinkMetrics.registerQueueGauges(queue);
+
+        this.workerThread = new Thread(this::run, "otlp-sink-buffer-thread");
+    }
+
+    private int getQueueCapacity() {
+        return Math.max(maxEvents * SAFETY_FACTOR, MIN_QUEUE_CAPACITY);
+    }
+
+    public boolean isRunning() {
+        return running && workerThread.isAlive();
+    }
+
+    public void start() {
+        running = true;
+        workerThread.start();
+    }
+
+    public void stop() {
+        running = false;
+        workerThread.interrupt();
+    }
+
+    /**
+     * Enqueues a span record for later batching and sending.
+     * <p>
+     * This will block if the internal queue is full, guaranteeing
+     * lossless delivery. On interruption, the span is rejected and
+     * error metrics are incremented.
+     *
+     * @param record the span record to enqueue
+     */
+    public void add(final Record<Span> record) {
+        try {
+            queue.put(record); // block until space available; guaranteeing lossless delivery
+        } catch (final InterruptedException e) {
+            Thread.currentThread().interrupt();
+            LOG.error("Interrupted while enqueuing span", e);
+            sinkMetrics.incrementRejectedSpansCount(1);
+            sinkMetrics.incrementErrorsCount();
+        }
+    }
+
+    /**
+     * Worker loop that batches spans by count, size, or time and then flushes them.
+     * <p>
+     * Continues running as long as {@link #running} is true or the queue is not empty.
+     * Handles encoding failures, timeout-based flush, and final flush on shutdown.
+     */
+    private void run() {
```

Comment:
Is there a particular reason you added a single thread for sending the data? The calls to `output` will run on a thread that you can use when the batch size is reached. This is how the other sinks work as well.

Reply:
@huypham612 , Data Prepper should be doing this already. We should double check if conditional routing somehow broke this. But it would call the `sink` for every processing iteration even if the events were empty. Other sinks expect this as well.